File: webapp/src/main/vueapp/src/logic/models/Doctor.ts
import {GenericEntity} from '~/logic/models/utils/GenericEntity';
import {Nullable} from '~/logic/Utils';
import {Office} from '~/logic/models/Office';
import {User} from '~/logic/models/User';

export class Doctor extends GenericEntity<Doctor> {
    private _id: number;
    private _phone: Nullable<string>;
    private _email: string;
    private _registrationNumber: Nullable<number>;
    private _user: User;
    private _office: Office;
    private _specialtyIds: Array<number>;

    public get id(): number {
        return this._id;
    }

    public set id(value: number) {
        this._id = value;
    }

    public get email(): string {
        return this._email;
    }

    public set email(value: string) {
        this._email = value;
    }

    public get phone(): Nullable<string> {
        return this._phone;
    }

    public set phone(value: Nullable<string>) {
        this._phone = value;
    }

    public get registrationNumber(): Nullable<number> {
        return this._registrationNumber;
    }

    public set registrationNumber(value: Nullable<number>) {
        this._registrationNumber = value;
    }

    public get user(): User {
        return this._user;
    }

    public set user(value: User) {
        this._user = value;
    }

    public get office(): Office {
        return this._office;
    }

    public set office(value: Office) {
        this._office = value;
    }

    public get specialtyIds(): Array<number> {
        return this._specialtyIds;
    }

    public set specialtyIds(value: Array<number>) {
        this._specialtyIds = value;
    }
}

File: models/src/main/java/ar/edu/itba/paw/models/Doctor.java
package ar.edu.itba.paw.models;

import javax.persistence.*;
import java.util.Collection;
import java.util.LinkedList;

@Entity
@Table(
        name = "doctor",
        indexes = {
                @Index(columnList = "doctor_id", name = "doctor_doctor_id_uindex", unique = true),
                @Index(columnList = "user_id", name = "doctor_user_id_index")
        }
)
public class Doctor extends GenericModel<Integer> {
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "doctor_doctor_id_seq")
    @SequenceGenerator(sequenceName = "doctor_doctor_id_seq", name = "doctor_doctor_id_seq", allocationSize = 1)
    @Column(name = "doctor_id")
    private Integer id;
    @Column(name = "phone")
    private String phone;
    @Column(name = "email")
    private String email;
    @Column(name = "registration_number")
    private Integer registrationNumber;
    @JoinColumn(name = "user_id", nullable = false)
    @ManyToOne(fetch = FetchType.EAGER)
    private User user;
    @JoinColumn(name = "office_id", nullable = false)
    @ManyToOne(fetch = FetchType.EAGER)
    private Office office;
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
            name = "system_doctor_specialty_doctor",
            joinColumns = @JoinColumn(name = "doctor_id"),
            inverseJoinColumns = @JoinColumn(name = "specialty_id")
    )
    @OrderBy("name ASC")
    private Collection<DoctorSpecialty> doctorSpecialties = new LinkedList<>();

    @Override
    public Integer getId() {
        return this.id;
    }

    @Override
    public void setId(Integer id) {
        this.id = id;
    }

    public String getPhone() {
        return this.phone;
    }

    public void setPhone(String phone) {
        this.phone = phone;
    }

    public String getEmail() {
        return this.email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public int getRegistrationNumber() {
        return this.registrationNumber == null ? 0 : this.registrationNumber;
    }

    public void setRegistrationNumber(Integer registrationNumber) {
        this.registrationNumber = registrationNumber;
    }

    public Office getOffice() {
        return this.office;
    }

    public User getUser() {
        return this.user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public void setOffice(Office office) {
        this.office = office;
    }

    public Collection<DoctorSpecialty> getDoctorSpecialties() {
        return this.doctorSpecialties;
    }

    public void setDoctorSpecialties(Collection<DoctorSpecialty> doctorSpecialties) {
        this.doctorSpecialties = doctorSpecialties;
    }

    @Override
    protected boolean isSameType(Object o) {
        return o instanceof Doctor;
    }
}
